[TensorExpr] Disallow arithmetic binary operations on Bool

asuhan opened the pull request:
Summary:

Arithmetic operations on Bool aren't fully supported in the evaluator. Moreover,
such semantics can be implemented by the client code through insertion of
explicit casts to widen and narrow to the desired types.

Test Plan:

test_tensorexpr --gtest_filter=TensorExprTest.ExprDisallowBoolArithmetic
python test/test_jit_fuser_te.py

asuhan committed:
[TensorExpr] Disallow arithmetic binary operations on Bool

Summary:

Arithmetic operations on Bool aren't fully supported in the evaluator. Moreover,
such semantics can be implemented by the client code through insertion of
explicit casts to widen and narrow to the desired types.

Test Plan:

test_tensorexpr --gtest_filter=TensorExprTest.ExprDisallowBoolArithmetic

Reviewers:

Subscribers:

Tasks:

Tags:

nickgg reviewed (approved):
Sure, assuming tests pass

bertmaher reviewed (changes requested):
I’m fairly certain some of these, at least the bit wise operators, are valid and can actually be generated from userland python code...

asuhan commented:
@bertmaher The bitwise operations don't work with the IR evaluator, I've checked.

bertmaher reviewed (commented):
I think we need to find out what eager mode does with these operators and what JIT IR is constructed when such a thing occurs.  Check out test_unary_ops in test_jit_fuser_te for an example of the kind of test we need to have on binary ops.  

If we implement these ops they need to be implemented the same way as eager mode, and if we don’t implement them (but eager does) we need to not form fusion groups containing them (check out tensorexpr_fuser.cpp) or we’ll break pytorch semantics.  If eager mode doesn’t implement them either, then of course we can also reject them.

asuhan commented:
@bertmaher I agree we must replicate eager semantics. To the extent we have test coverage for them, I'm fixing them - I just did so for `TestTEFuser.test_clamp` and it passes now.

Do you think we have usage lacking test coverage which also happen to not use `IRSimplifier`? Because I believe this patch could only add additional breakage on those paths. Any path which relies on arithmetic operations being supported for boolean would break when calling `IRSimplifier::simplify`.

bertmaher commented:
I would be entirely unsurprised that we have broken code paths here already exposed to users :-(.  What I'm arguing is that instead of making a change that breaks these untested code paths in a different way, that we want to add tests for those code paths and do whatever needs done to make them work properly.

asuhan commented:
@bertmaher That's reasonable, I agree that we must clean up all these paths which would break under implicit casts and promotion of boolean to numeric. Given your approval of https://github.com/pytorch/pytorch/pull/44659 as-is, I'm fine with landing that first, unblocking myself and then we can fix the current brittleness. Does this sound good to you?

bertmaher commented:
Sounds great, thank you!

asuhan commented:
@bertmaher I've addressed issues and added tests for all binary operations which are affected by this. A few things to note:

* `sub`, `__lshift__` and `__rshift__` don't support boolean in eager, therefore we don't need to do anything about their inputs.
* `max` and `min` are already broken for input types with the CUDA back-end, because we're calling `isnan`, which is ambiguous for integer inputs. I plan to address this in a separate PR.

Please take a look.

bertmaher commented on test/cpp/tensorexpr/test_expr.cpp:
```diff
@@ -164,6 +164,24 @@ void testExprDoubleTest() {
+  VarHandle x("x", kBool);
+  VarHandle y("y", kBool);
+  std::string error{"arithmetic binary operations on Bool not supported"};
+  ASSERT_THROWS_WITH((x + y), error);
+  ASSERT_THROWS_WITH((x - y), error);
```
If I understand, you're basically saying bool arithmetic within the TE IR should be disallowed?  Sounds good to me.

asuhan replied:
@bertmaher Yes, which addresses the testing debt introduced by https://github.com/pytorch/pytorch/pull/44659. By outright disallowing them, it's one less thing to worry about and write tests for.

bertmaher reviewed (approved):
Thanks you for cleaning this up and adding a great test!
